In [1]:
# Import libraries 
import pandas as pd
import numpy as np
import os

In [3]:
# Create path to project folder
path = r'C:\Users\tatal\06-2023 Instacart Basket Analysis'

## From the exercise

# Dowload new df_ords_prior dataset and import it to Jupiter 

In [5]:
## Import the new df_ords_prior from Orignal folder
df_ords_prior = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'order_products__prior.csv'), index_col = False)

## Import Order dataset from Prepared folder

In [8]:
## Import the checked versiono of order from Prepared folder
df_order_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'order_checked.csv'), index_col = [0])

## Check Both dataset with head

In [11]:
df_ords_prior.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [9]:
df_order_checked.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

## run shape to understand the data

In [15]:
df_ords_prior.shape

(32434489, 4)

In [16]:
df_order_checked.shape

(3421083, 6)

## quite large database as it has all items in the orders. We want to merge this new dataframe with the order dataframe by order_id column

In [18]:
## Inner join the 2 tables by order id
df_merged_large = df_order_checked.merge(df_ords_prior, on = 'order_id', indicator = True)

In [19]:
## Check new merged file
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered _merge  
0                    NaN         196                  1          0   both  
1                    NaN       14084                  2          0   both  
2                    NaN       12427                  3          0   both  
3                    NaN       26088                  4          0   both  
4                    NaN       26405                  5          0   both

In [20]:
## Count merge to see if it exist in both
df_merged_large['_merge'].value_counts()

both          32434489
left_only            0
right_only           0
Name: _merge, dtype: int64

##all order id exist in both dataset when using inner join. Becaue it only took the common lines between both. The resultis dataframe (after the merge) has 32,434,489 rows, and each of those rows have information found in both input data sets. Using the outer merge (how = 'outer') shows that are more than 20K row missing on new dataset (let only), but for this project we only need to consider the common rows. So using inner join is correct

## Exporting Data in Pickle instead cvs 

In [21]:
#export the merge data as pickle as it's too large
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))